implementations/rust/ockam/ockam_identity/tests/message_flow_auth.rs

```diff
@@ -0,0 +1,316 @@
+            .first()
+            .unwrap()
+            .encryptor_messaging_address()
+            .clone()
+    }
+}
+
+async fn create_secure_channel_listener(
+    ctx: &Context,
+    session: &Option<(Sessions, SessionId)>,
+) -> Result<SecureChannelListenerInfo> {
+    let identity = Identity::create(ctx, &Vault::create()).await?;
+
+    let trust_options =
+        SecureChannelListenerTrustOptions::new().with_trust_policy(TrustEveryonePolicy);
+    let trust_options = if let Some((sessions, session_id)) = session {
+        trust_options.with_session(sessions, session_id)
+    } else {
+        trust_options
+    };
+
+    identity
+        .create_secure_channel_listener("listener", trust_options)
+        .await?;
+
+    let info = SecureChannelListenerInfo { identity };
+
+    Ok(info)
+}
+
+struct SecureChannelInfo {
+    identity: Identity<Vault, InMemoryStorage>,
+    address: Address,
+}
+
+async fn create_secure_channel(
+    ctx: &Context,
+    connection: &TcpConnectionInfo,
+) -> Result<SecureChannelInfo> {
+    let identity = Identity::create(ctx, &Vault::create()).await?;
+
+    let trust_options = SecureChannelTrustOptions::new();
+    let trust_options = if let Some((sessions, session_id)) = &connection.session {
+        trust_options.with_ciphertext_session(sessions, session_id)
+    } else {
+        trust_options
+    };
+
+    let address = identity
+        .create_secure_channel(
+            route![connection.address.clone(), "listener"],
+            trust_options,
+        )
+        .await?;
+
+    let info = SecureChannelInfo { identity, address };
+
+    Ok(info)
+}
+
+#[allow(non_snake_case)]
+#[ockam_macros::test]
+async fn sessions__secure_channel_over_tcp_without_session__should_pass_messages(
+    ctx: &mut Context,
+) -> Result<()> {
+    let bob_tcp_info = create_tcp_listener(ctx, false).await?;
+
+    let bob_listener_info = create_secure_channel_listener(ctx, &bob_tcp_info.session).await?;
+
+    let connection_to_bob = create_connection(ctx, &bob_tcp_info.socket_addr, false).await?;
+    ctx.sleep(Duration::from_millis(50)).await; // Wait for workers to add themselves to the registry
+    let connection_to_alice = bob_tcp_info.get_connection();
+
+    check_message_flow(ctx, route![connection_to_bob.address.clone()], true).await?;
+    check_message_flow(ctx, route![connection_to_alice.clone()], true).await?;
+
+    let channel_to_bob = create_secure_channel(ctx, &connection_to_bob).await?;
+    ctx.sleep(Duration::from_millis(50)).await; // Wait for workers to add themselves to the registry
+    let channel_to_alice = bob_listener_info.get_channel();
+
+    check_message_flow(ctx, route![channel_to_bob.address.clone()], true).await?;
+    check_message_flow(ctx, route![channel_to_alice.clone()], true).await?;
+
+    ctx.stop().await
+}
+
+#[allow(non_snake_case)]
+#[ockam_macros::test]
+async fn sessions__secure_channel_over_tcp_with_alice_session__should_not_pass_messages(
```

Comment:
The naming of the tests does not make it very obvious what we are testing and if we are testing all the possible combinations. Maybe you can try to refactor the code so that there is function taking: 

 - 4 input parameters
    - `tcp_listener_creation`: `WithSession/WithoutSession`
    - `tcp_connection_creation`: `WithSession/WithoutSession`
    - `secure_channel_listener_creation`: `WithSession/WithoutSession`
    - `secure_channel_creation`: `WithSession/WithoutSession`

 - 4 expected results
    - `outgoing_message_from_bob_tcp_receiver`: `Received/NotReceived`
    - `outgoing_message_from_bob_tcp_sender`: `Received/NotReceived` 
    - `outgoing_message_from_bob_encryptor`: `Received/NotReceived`
    - `outgoing_message_from_bob_decryptor`: `Received/NotReceived` 

What I wrote above is probably incorrect and makes me think that we should make a first separation between the listening and the connecting side and test:

 1. the topology of a TcpListener + SecureChannel listener (bob), with and without a session
       - can a message be sent from the `TcpRecvProcessor` when a connection is created?
       - can a message be sent from the `TcpSender` when a connection is created?
       - can a message be sent from the `Decryptor` (via the the `TcpRcvProcessor`) when a secure channel is created?
       - can a message be sent from the `Encryptor` (via the the `TcpSender`) when a secure channel is created?

 2. the topology of a TcpConnection + SecureChannel connection (alice), with and without a session
       -  similar tests here

Reply:
I don't think it's worth refactoring it even more right now

Reply:
Ok, but I'm not sure we cover all the cases we should be covering there. Right now I don't see how the coverage is systematic and exhaustive. Maybe you can cover this in a follow-up PR, or convince me that we're good :-)